### PR TITLE
#25790 Updating type hints to Python3 syntax in pandas/core/array

### PR DESCRIPTION
**Pull request by gwrome**
- [x] addresses the other half of #25790
- [x] pandas/tests/arrays/ tests passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/core/arrays/array_.py [hunk 1]
```diff
@@ -11,11 +11,14 @@
 from pandas import compat
 
 
-def array(data,         # type: Sequence[object]
-          dtype=None,   # type: Optional[Union[str, np.dtype, ExtensionDtype]]
-          copy=True,    # type: bool
-          ):
-    # type: (...) -> ExtensionArray
+if TYPE_CHECKING:
```
Comment (jreback) [inline pandas/core/arrays/array_.py]: why is this needed? conditional imports for type checking are not supportable

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/array_.py (hunk 1 above)
Comment (gwrome) [reply pandas/core/arrays/array_.py]: As I understand it, we need ExtensionArray to be defined to check this function's return type. Since ExtensionArray is defined elsewhere, we need to import it during type checking. I based this code on the approach described in [PEP 484 — Runtime or type checking](https://www.python.org/dev/peps/pep-0484/#runtime-or-type-checking).

Is there  a better, supportable way to do this?

Comment (WillAyd) [reply pandas/core/arrays/array_.py]: Yea this is the same conversation as https://github.com/pandas-dev/pandas/pull/25802#discussion_r267857645

@jreback our support of 3.5 includes 3.5.0 and 3.5.1 right? This wasn't introduced until 3.5.2

Comment (jreback) [reply pandas/core/arrays/array_.py]: I would instead simply use the ABCExtensionArray classes which would work for this (and lots of other clases). we should maybe just do this generally. These are by-definition available with no dependencies for this exact purpose.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/base.py [hunk 2]
```diff
@@ -343,8 +337,7 @@ def astype(self, dtype, copy=True):
         """
         return np.array(self, dtype=dtype, copy=copy)
 
-    def isna(self):
-        # type: () -> Union[ExtensionArray, np.ndarray]
+    def isna(self) -> Union['ExtensionArray', np.ndarray]:
```
Comment (WillAyd) [inline pandas/core/arrays/base.py]: Might as well use the ABC here and throughout module

File: pandas/core/arrays/datetimelike.py [hunk 3]
```diff
@@ -58,8 +58,7 @@ def _get_attributes_dict(self):
         return {k: getattr(self, k, None) for k in self._attributes}
 
     @property
-    def _scalar_type(self):
-        # type: () -> Union[type, Tuple[type]]
+    def _scalar_type(self) -> Union[type, Tuple[type]]:
```
Comment (WillAyd) [inline pandas/core/arrays/datetimelike.py]: Let's use Type consistently

File: pandas/core/arrays/period.py [hunk 4]
```diff
@@ -183,8 +183,12 @@ def _simple_new(cls, values, freq=None, **kwargs):
...
 
     @classmethod
-    def _from_sequence(cls, scalars, dtype=None, copy=False):
-        # type: (Sequence[Optional[Period]], PeriodDtype, bool) -> PeriodArray
+    def _from_sequence(
+            cls,
+            scalars: Sequence[Optional[Period]],
+            dtype: PeriodDtype = None,
+            copy: bool = False,
+    ) -> 'PeriodArray':
```
Comment (WillAyd) [inline pandas/core/arrays/period.py]: ABCPeriodArray

File: pandas/core/arrays/sparse.py [hunk 5]
```diff
@@ -397,8 +399,12 @@ def _get_fill(arr):
         return np.asarray(arr.fill_value)
 
 
-def _sparse_array_op(left, right, op, name):
-    # type: (SparseArray, SparseArray, Callable, str) -> Any
+def _sparse_array_op(
+        left: 'SparseArray',
```
Comment (WillAyd) [inline pandas/core/arrays/sparse.py]: ABCSparseArray

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/sparse.py (hunk 5 above)
Comment (gwrome) [reply pandas/core/arrays/sparse.py]: I think we're good to go @WillAyd @jreback

File: pandas/core/arrays/sparse.py [hunk 6]
```diff
@@ -27,7 +27,7 @@
     pandas_dtype)
 from pandas.core.dtypes.dtypes import register_extension_dtype
 from pandas.core.dtypes.generic import (
-    ABCIndexClass, ABCSeries, ABCSparseSeries)
+    ABCIndexClass, ABCSeries, ABCSparseArray, ABCSparseSeries)
```
Comment (WillAyd) [inline pandas/core/arrays/sparse.py]: I think you want to use this where you currently have forward declarations too?

**Review by WillAyd — COMMENTED** [review]
@gwrome one other thing that could be helpful - since some if not all of these modules are blacklisted from mypy could you manually run locally for the modules and compare this PR to head? Just want to make sure there are no new regressions

File: pandas/core/arrays/sparse.py (hunk 6 above)
Comment (gwrome) [reply pandas/core/arrays/sparse.py]: Ran mypy against the changed files without the blacklist, and there aren't any new problems.

File: pandas/core/arrays/period.py [hunk 7]
```diff
@@ -540,14 +542,9 @@ def _sub_period(self, other):
...
             self,
-            other,      # type: Union[ExtensionArray, np.ndarray[int]]
-            op          # type: Callable[Any, Any]
-    ):
-        # type: (...) -> PeriodArray
-
-        # TODO: ABCIndexClass is a valid type for other but had to be excluded
-        # due to length of Py2 compatability comment; add back in once migrated
-        # to Py3 syntax
+            other: Union[ExtensionArray, np.ndarray, ABCIndexClass],
```
Comment (WillAyd) [inline pandas/core/arrays/period.py]: Any reason for removing the type for ndarray?

File: pandas/core/arrays/sparse.py [hunk 8]
```diff
@@ -77,8 +77,11 @@ class SparseDtype(ExtensionDtype):
     # hash(nan) is (sometimes?) 0.
     _metadata = ('_dtype', '_fill_value', '_is_na_fill_value')
 
-    def __init__(self, dtype=np.float64, fill_value=None):
-        # type: (Union[str, np.dtype, 'ExtensionDtype', type], Any) -> None
+    def __init__(
+            self,
+            dtype: Union[str, np.dtype, 'ExtensionDtype', type] = np.float64,
```
Comment (WillAyd) [inline pandas/core/arrays/sparse.py]: Let's just always use Type

**Review by WillAyd — CHANGES_REQUESTED** [review]
Couple minor things. IMO this is close - obviously some of these annotations are wrong and may need to be fixed but can be done in subsequent PRs

File: pandas/core/arrays/period.py (hunk 7 above)
Comment (gwrome) [reply pandas/core/arrays/period.py]: During my testing at the time (before you added mypy.ini and the rest of the type checking framework to the project), mypy was throwing an error that suggested that ndarrays couldn't take a subscript, so I dropped it. I probably had something in my testing environment set up incorrectly (or maybe more strictly, not sure), because it doesn't throw that error with the current mypy.ini minus the blacklist. Will add it back.

Comment (gwrome) [reply pandas/core/arrays/period.py]: Wait, it wasn't mypy. `code_checks doctests` throws 
```
...pandas/core/arrays/period.py", line 545, in PeriodArray
    other: Union[ExtensionArray, np.ndarray[int], ABCIndexClass],
TypeError: 'type' object is not subscriptable
```

File: pandas/core/arrays/base.py [hunk 9]
```diff
@@ -343,8 +338,7 @@ def astype(self, dtype, copy=True):
         """
         return np.array(self, dtype=dtype, copy=copy)
 
-    def isna(self):
-        # type: () -> Union[ExtensionArray, np.ndarray]
+    def isna(self) -> Union[ABCExtensionArray, np.ndarray]:
```
Comment (jreback) [inline pandas/core/arrays/base.py]: so this is a the first stab at ArrayLike (e.g. to put into pandas.typing)

File: pandas/core/arrays/sparse.py [hunk 10]
```diff
@@ -77,8 +77,11 @@ class SparseDtype(ExtensionDtype):
     # hash(nan) is (sometimes?) 0.
     _metadata = ('_dtype', '_fill_value', '_is_na_fill_value')
 
-    def __init__(self, dtype=np.float64, fill_value=None):
-        # type: (Union[str, np.dtype, 'ExtensionDtype', type], Any) -> None
+    def __init__(
+            self,
+            dtype: Union[str, np.dtype, ExtensionDtype, Type] = np.float64,
```
Comment (jreback) [inline pandas/core/arrays/sparse.py]: this could be Dtype (in pandas.typing)

File: pandas/core/arrays/base.py (hunk 9 above)
Comment (gwrome) [reply pandas/core/arrays/base.py]: Seems like a reasonable place to start.

Comment (jreback) [reply pandas/core/arrays/base.py]: these are ok here for now, mainly adding things that we *should* add

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @WillAyd merge when ready

File: pandas/core/arrays/base.py (hunk 9 above)
Comment (WillAyd) [reply pandas/core/arrays/base.py]: Agreed this would be great @gwrome do you mind opening up an issue as a follow up?

File: pandas/core/arrays/period.py (hunk 7 above)
Comment (WillAyd) [reply pandas/core/arrays/period.py]: Gotcha. Kind of strange this throws an error now but not before. Something to revisit later

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
nice job @gwrome - thanks for helping to kick off these typing initiatives

File: pandas/core/arrays/base.py (hunk 9 above)
Comment (gwrome) [reply pandas/core/arrays/base.py]: No problem. I should be able to do that this weekend.
